Allow better caching of SPA launch pages

rtibbles opened the pull request:
### Summary
Sets up language URL prefixing to allow SPA launch pages to be cached with a minimal number of Vary headers set.

Sets up a 300 second cache on all SPA launch pages.

### Reviewer guidance
Does everything work as expected? Do redirects work well?

### Reference
Dependent on #4984.

----

### Contributor Checklist


PR process:

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

Testing:

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

### Reviewer Checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.rst

rtibbles committed:
Remove user specific data from templates.
Add experimental no-user-data cache decorator.

rtibbles committed:
Add separate option for configuring api_urls in plugins.

rtibbles committed:
Make urls.py and api_urls.py in plugins auto-detected by default.

rtibbles committed:
Put all view URLs under language prefixing.
Add tests for behaviour.

rtibbles committed:
Set up additional caching behaviour.

rtibbles committed:
Prevent language redirect on API views.
Fix redirect tests.

indirectlylit commented on kolibri/core/decorators.py:
```diff
@@ -275,3 +276,19 @@ def wrapped_view(*args, **kwargs):
+        response = view_func(*args, **kwargs)
+        headers = ['accept-encoding', 'accept', 'user-agent']
+        patch_response_headers(response, cache_timeout=300)
+        response['Vary'] = ', '.join(headers)
```
could be simplified, unless the intent was to merge these headers with some existing headers?

```python
response['Vary'] = 'accept-encoding, accept, user-agent'
```

indirectlylit commented on kolibri/core/decorators.py:
```diff
@@ -275,3 +276,19 @@ def wrapped_view(*args, **kwargs):
+    """
+    Set appropriate Vary on headers on a view that specify there is
+    no user specific data being rendered in the view.
+    """
```
how does this interact with bootstrapped data and authorization to access certain pages and data?

rtibbles replied:
This shouldn't be used on pages that bootstrap user data into the page. I have also removed most instances of bootstrapped user data in the page, because it would interfere with this.

indirectlylit reviewed (commented):
Looks good from a cursory review.

Do you have a sense for what kind of performance benefits this achieves?

I'm a bit concerned that this touches so many files so close to 0.12.0, and particularly stuff that's harder to test like i18n and caching behaviors.

rtibbles commented:
We can hold off on this until 0.12.1 - I mostly want to have a few things in our pocket to allow performance improvements if we discover that we need them for 0.12.0.

indirectlylit commented:
setting milestone to 0.12.1

note there's a conflict in kolibri_app.js

rtibbles commented:
Not a hugely representative methodology, but I ran siege with default settings (25 concurrent users), and these are the results on current release-v0.12.x:
```
Transactions:		        1000 hits
Availability:		      100.00 %
Elapsed time:		       65.42 secs
Data transferred:	        6.90 MB
Response time:		        1.60 secs
Transaction rate:	       15.29 trans/sec
Throughput:		        0.11 MB/sec
Concurrency:		       24.53
Successful transactions:        1000
Failed transactions:	           0
Longest transaction:	        5.91
Shortest transaction:	        0.18
```

These are the results when run on this PR:
```
Transactions:		        1000 hits
Availability:		      100.00 %
Elapsed time:		       10.00 secs
Data transferred:	        6.76 MB
Response time:		        0.25 secs
Transaction rate:	      100.00 trans/sec
Throughput:		        0.68 MB/sec
Concurrency:		       24.72
Successful transactions:        1000
Failed transactions:	           0
Longest transaction:	        3.69
Shortest transaction:	        0.06
```

Testing conditions definitely favour this PR and what it is doing, but it is still a roughly 6 times speed up on average response time.

rtibbles commented:
The other thing to note, obviously, is that this is moving some bootstrapping out of user pages, so this will result in more requests to the server, so we would probably want to do a live load test to properly assess the impact of this.

benjaoming commented on kolibri/core/decorators.py:
```diff
@@ -275,3 +276,19 @@ def wrapped_view(*args, **kwargs):
+    """
+    def inner_func(*args, **kwargs):
+        request = args[0]
+        del request.session
```
Do we want a decorator that just deletes the entire session data? Where will this be used? Normally, it would seem like a logout procedure. This doesn't match that well with the `__docstring__` of the decorator.

jredrejo replied:
I understand the utility of this decorator except for the point Ben comments, why is it doing `del request.session`?
All the other lines in the decorator make sense for me, but I don't understand the rationale of deleting the session when switching between learn/learn.html and user/user.html

rtibbles replied:
The reason for deleting the session is that otherwise Django adds a Vary header for cookies onto the request, meaning no caching takes place.

I could achieve the same effect by having this decorator add a "don't add cookie vary header" flag to the view, and then modify other session modifying middleware to obey that directive, but this seemed simplest.

benjaoming commented on kolibri/core/device/middleware.py:
```diff
@@ -1,17 +1,70 @@
 
 
-class KolibriLocaleMiddleware(LocaleMiddleware):
+class KolibriLocaleMiddleware:
```
Why was the inheritance removed? It would be okay to override all the current methods, but keeping the inheritance somehow also means keeping the contract of what a `LocaleMiddleware` is supposed to do.

jredrejo replied:
@benjaoming I think the reasons are explained in the related PR #4984 :

> Also cleans up some unused code (including the sessions middleware which is a no op in Django - the presence of it in settings is only used to trigger some backwards compatibility behaviour that we do not need).

benjaoming replied:
I think it's better to keep it and override changed/unneeded methods. Simply because of maintaining the link to the reference implementation.

rtibbles replied:
The other reason for not inheriting is that the default Django locale middleware class is an old style middleware, so I would be overriding its methods simply to turn them into a no op (and I am not entirely sure how using new style middleware works when inheriting from the middleware deprecation compatibility class).

jredrejo commented on kolibri/core/context_processors/custom_context_processor.py:
```diff
@@ -33,12 +33,10 @@ def pass_browser_entry(agent, entry):
-    return {'supported_browser': request.session['supported_browser']}
+    user_agent = user_agents.parse(request.META.get('HTTP_USER_AGENT', ''))
+    return {
+        'supported_browser': all(pass_browser_entry(user_agent, entry) for entry in browser_requirements)
```
I see the returned value is being removed from the session. 
wouldn't it be good to use the session to cache the value and return it if it already exixts? it's being called/recalculated from the base.html template whenever the SPA changes.

rtibbles replied:
I can either cache the pages, or I can use the session object. A middle ground might be to memoize the calculation based on user agent?

jredrejo replied:
Yeah, I hadn't thought of using the user agent, but your idea of memoizing based on it would be the more efficient solution.

jredrejo commented on kolibri/deployment/default/settings/dev.py:
```diff
@@ -11,3 +11,10 @@
 
 DEVELOPER_MODE = True
+
+# Create a dummy cache for each cache
```
thus, nulls the caches, right?

rtibbles replied:
Yes, I am open to arguments that we should enforce caching in development to expose caching induced bugs to devs, but this will prevent having to do superfluous saves on python files to restart the devserver just to clear the cache.

jredrejo replied:
nah, I just wanted to confirm I had understood your purpose correctly.

rtibbles committed:
Add cherrypy caching for all served routes, except content.

indirectlylit commented:
switching base to `develop`

rtibbles commented:
This PR is now dependent on #5246

rtibbles committed:
Do not set any translation mechanisms on requests
unless for a translated view.

rtibbles commented:
Note that a live load test of this showed improved performance of this over current release 0.12 - although there was a cache warm-up phase where things were a little slower.

rtibbles committed:
Address comments.

rtibbles committed:
Remove gzip middleware to increase API performance.

rtibbles committed:
Move all user agent specific code the frontend.
Remove user agent vary header on SPA pages.

rtibbles committed:
Remove unused test.

rtibbles commented:
This PR now also removes the GZip middleware which seems to have been a considerable performance bottleneck on our bulkier API endpoints. As an example, here is a siege run against the channel endpoint, with and without GZip:
```
Channel Endpoint:
With GZip:
Transactions:		         500 hits
Availability:		      100.00 %
Elapsed time:		       10.04 secs
Data transferred:	       80.57 MB
Response time:		        0.49 secs
Transaction rate:	       49.80 trans/sec
Throughput:		        8.03 MB/sec
Concurrency:		       24.63
Successful transactions:         500
Failed transactions:	           0
Longest transaction:	        2.27
Shortest transaction:	        0.08

Without GZip:
Transactions:		         500 hits
Availability:		      100.00 %
Elapsed time:		        5.79 secs
Data transferred:	      111.32 MB
Response time:		        0.24 secs
Transaction rate:	       86.36 trans/sec
Throughput:		       19.23 MB/sec
Concurrency:		       20.75
Successful transactions:         500
Failed transactions:	           0
Longest transaction:	        3.02
Shortest transaction:	        0.01
```

rtibbles committed:
Move pskolibri to kolibri.utils

rtibbles committed:
Add dynamic cache size calculation.

indirectlylit commented on kolibri/core/device/translation.py:
```diff
@@ -3,58 +3,91 @@
+    try:
+        # If this is not a view that needs to be translated, return None, and be done with it!
+        if not getattr(resolve(request.path_info).func, "translated", False):
+            return None, False
```
technically, I think this should be `None, True`

rtibbles replied:
`The language has been calculated from the path.`

I am using a Russellian theory of descriptions here to return `False`, because of the null referent of `language` (`language` is `None`), the predicate evaluates to `False`, just as `The language has not been calculated from the path` would evaluate to `False` by the same reasoning.

indirectlylit replied:
my cat's breath smells like cat food

indirectlylit commented on kolibri/core/device/middleware.py:
```diff
@@ -1,34 +1,62 @@
     def __call__(self, request):
-        # Code to be executed for each request before
-        # the view (and later middleware) are called.
+        language, language_from_path = get_language_from_request(request)
```
maybe rename to `get_language_and_reason(request)` or something to convey multiple return values?

rtibbles replied:
At the very least I should update the doc string of the function - coming up with a concise and meaningful name might be trickier.

indirectlylit commented on kolibri/core/assets/src/utils/userAgent.js:
```diff
@@ -0,0 +1,116 @@
+  return browser;
+}
+
+export function passesRequirements(browser, requirements) {
```
recommend adding a brief comment about the expected content of `browser` and `requirements`

rtibbles replied:
Yeah, JSDocs needed.

indirectlylit commented on kolibri/core/decorators.py:
```diff
@@ -320,3 +321,26 @@ def wrapped_view(*args, **kwargs):
+        del request.session
+        response = view_func(*args, **kwargs)
+        patch_response_headers(response, cache_timeout=300)
+        response["Vary"] = "accept-encoding, accept"
```
interesting. this is the minimal case?

rtibbles replied:
I believe so, not varying on these could cause problems if we sent the wrong mimetype back to a request, or tried to compress for a client that couldn't handle compression.

`accept-charset` is another possibility, but I have never seen our backend add `accept-charset` to the Vary header. My hope is this would be subsumed by language?

indirectlylit commented on kolibri/core/device/middleware.py:
```diff
@@ -1,34 +1,62 @@
+    """
+    Copied and then modified into a new style middleware from:
+    https://github.com/django/django/blob/stable/1.11.x/django/middleware/locale.py
+    Also has several other changes to suit our purposes.
```
maybe include a description of what the intended logic is, especially with regard to redirects? It's a bit hard to evaluate otherwise

indirectlylit commented on kolibri/core/device/middleware.py:
```diff
@@ -1,34 +1,62 @@
+                    return HttpResponseRedirect(language_url)
+
+            if "Content-Language" not in response:
+                response["Content-Language"] = language
```
Since this block has some complex nested if/else statements, recommend adding a few more comments to conditional branches indicating the circumstances under which you'd expect the code to reach the branch.

Another thing that can be helpful is to stick `return` statements anywhere the processing is supposed to be "done". This can help reduce ambiguity of intent compared to letting execution fall out of the conditionals to the final `return`.

rtibbles replied:
> Another thing that can be helpful is to stick return statements anywhere the processing is supposed to be "done". This can help reduce ambiguity of intent compared to letting execution fall out of the conditionals to the final `return`.

This is already the case, the return of the redirect is the shortcut return.

indirectlylit commented on kolibri/utils/server.py:
```diff
@@ -188,6 +194,33 @@ def stop(pid=None, force=False):
+        return MAX_CACHE
+    return MIN_CACHE
+
+
```
this looks familiar :)

rtibbles replied:
Only done it twice so far, so just changed a few names to protect the innocent.

indirectlylit reviewed (commented):
This is looking good - I think having languages segmented under their own URLs is a good move, and it makes sense to do it in support of better caching.

There are a lot of possible states in here, given conditional logic for language selection, browser compatibility, and page caching behaviors. Most of my notes relate to future maintainability in terms of clarifying intended behavior.

The test coverage is a great defense against edge cases bugs, but I haven't reviewed them carefully yet.

In terms of manual testing it will be hard to cover every scenario. I think it might be helpful to add some gherkin stories which exercise various paths through the conditionals, at least for the intended "happy paths".

I started testing briefly, but the language switching didn't seem to work on `devserver-hot`:

![switch](https://user-images.githubusercontent.com/2367265/55362231-feecab80-549e-11e9-9e3c-6556308972ff.gif)

rtibbles commented:
Thanks - I'll double check the language switching behaviour from the frontend some more, and see if I can replicate.

rtibbles committed:
Add documentation to and unabstract user agent code.

rtibbles committed:
Update tests for locale_middleware to check behaviour when the URL
PREFIX option is used.

rtibbles committed:
Add documentation and simplify implementation for readability.

rtibbles committed:
Update learn plugin references for simpler plugin name.

rtibbles committed:
Return a redirect URL to ensure proper redirect after change.

rtibbles committed:
Reimplement set lang redirect with URL translation.
Default to redirect user if not a valid local URL.

rtibbles commented:
Have now addressed all comments - the language changing behaviour was indeed broken. Have fixed the issue, and also reimplemented redirect to make sure that when you change language it returns you to exactly the same page (including fragment and query param) that you were on previously.

I think this is ready to merge, assuming there are no tests that I have missed.

indirectlylit reviewed (approved):
did a bit of testing, and everything seems in order. Let's get it in!

I think there's an open interesting question of "permalinks" and whether they should contain the target language or not...
